### PR TITLE
Add getAllWallets to BaseAccountFactory

### PR DESCRIPTION
**Pull request by nkrishang**

### CODE REVIEW

File: contracts/smart-wallet/dynamic/DynamicAccountFactory.sol [hunk 1]
```diff
@@ -27,7 +27,7 @@ contract DynamicAccountFactory is BaseAccountFactory, ContractMetadata, Permissi
     //////////////////////////////////////////////////////////////*/
 
     constructor(IEntryPoint _entrypoint, IExtension.Extension[] memory _defaultExtensions)
-        BaseAccountFactory(payable(address(new DynamicAccount(_entrypoint, _defaultExtensions))))
+        BaseAccountFactory(payable(address(new DynamicAccount(_entrypoint, _defaultExtensions))), address(_entrypoint))
```
Comment (joaquim-verges) [inline contracts/smart-wallet/dynamic/DynamicAccountFactory.sol:30]: why do you need the entrypoint address separately? can you not get it from the passed in `IEntrypoint` ?

File: contracts/smart-wallet/non-upgradeable/Account.sol [hunk 2]
```diff
@@ -205,6 +209,18 @@ contract Account is
...
 
+    /// @dev Registers the account on the factory if it hasn't been registered yet.
+    function _registerOnFactory() internal virtual {
+        AccountFactory factoryContract = AccountFactory(factory);
+        AccountStorage.Data storage data = AccountStorage.accountStorage();
+        if (!data.isRegistered) {
+            if (!factoryContract.isRegistered(address(this))) {
+                factoryContract.onRegister();
+            }
+            data.isRegistered = true;
```
Comment (joaquim-verges) [inline contracts/smart-wallet/non-upgradeable/Account.sol:220]: why do you need the account storage flag? isn't it sufficient to check the factory ? is it to save some gas?

File: contracts/smart-wallet/utils/AccountCore.sol [hunk 3]
```diff
@@ -62,6 +69,33 @@ contract AccountCore is IAccountCore, Initializable, Multicall, BaseAccount, ERC
         _setAdmin(_defaultAdmin, true);
     }
 
+    /*///////////////////////////////////////////////////////////////
+                            External functions
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Executes a transaction (called directly from an admin, or by entryPoint)
+    function execute(
```
Comment (joaquim-verges) [inline contracts/smart-wallet/utils/AccountCore.sol:77]: hmmm there's no way to keep this in the AccountExtension?

File: contracts/smart-wallet/utils/BaseAccountFactory.sol [hunk 4]
```diff
@@ -111,6 +124,16 @@ abstract contract BaseAccountFactory is IAccountFactory, Multicall {
                             View functions
     //////////////////////////////////////////////////////////////*/
 
+    /// @notice Returns whether an account is registered on this factory.
+    function isRegistered(address _account) external view returns (bool) {
+        return allAccounts.contains(_account);
+    }
+
+    /// @notice Returns all wallets created on the factory.
+    function getAllWallets() external view returns (address[] memory) {
```
Comment (joaquim-verges) [inline contracts/smart-wallet/utils/BaseAccountFactory.sol:133]: `getAllAccounts` to keep things consistent pls

File: contracts/lib/BytesLib.sol [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+                mstore(tempBytes, 0)
+
+                mstore(0x40, add(tempBytes, 0x20))
+            }
+        }
+
+        return tempBytes;
+    }
+
+    function toAddress(bytes memory _bytes, uint256 _start) internal pure returns (address) {
```
Comment (joaquim-verges) [inline contracts/lib/BytesLib.sol:285]: oooh thats neat. Do we need the entire lib ? I feel like we could just remove everything else but this and save some deployment gas

Comment (nkrishang) [reply contracts/lib/BytesLib.sol:15]: Removed everything other than `toAddress`

File: contracts/smart-wallet/dynamic/DynamicAccountFactory.sol (hunk 1 above)
Comment (nkrishang) [reply contracts/smart-wallet/dynamic/DynamicAccountFactory.sol:30]: Yep, `address(_entrypoint)` is indeed coming from the passed in `IEntryPoint _entrypoint`.

**Pull request merged**